server/handler/query.go

```diff
@@ -0,0 +1,192 @@
+package handler
+
+import (
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+
+	"github.com/src-d/gitbase-playground/server/serializer"
+	"gopkg.in/bblfsh/sdk.v1/uast"
+
+	"github.com/go-sql-driver/mysql"
+)
+
+type queryRequest struct {
+	Query string `json:"query"`
+	Limit int    `json:"limit"`
+}
+
+// genericVals returns a slice of interface{}, each one a pointer to the proper
+// type for each column
+func genericVals(colTypes []string) []interface{} {
+	columnValsPtr := make([]interface{}, len(colTypes))
+
+	for i, colType := range colTypes {
+		switch colType {
+		case "BIT":
+			columnValsPtr[i] = new(sql.NullBool)
+		case "TIMESTAMP", "DATE", "DATETIME":
+			columnValsPtr[i] = new(mysql.NullTime)
+		case "INT", "MEDIUMINT", "BIGINT", "SMALLINT", "TINYINT":
+			columnValsPtr[i] = new(sql.NullInt64)
+		case "DOUBLE", "FLOAT":
+			columnValsPtr[i] = new(sql.NullFloat64)
+		case "JSON":
+			columnValsPtr[i] = new([]byte)
+		default: // All the text and binary variations
+			columnValsPtr[i] = new(sql.NullString)
+		}
+	}
+
+	return columnValsPtr
+}
+
+// Query returns a function that forwards an SQL query to gitbase and returns
+// the rows as JSON
+func Query(db *sql.DB) RequestProcessFunc {
+	return func(r *http.Request) (*serializer.Response, error) {
+		var queryRequest queryRequest
+		body, err := ioutil.ReadAll(r.Body)
+		if err == nil {
+			err = json.Unmarshal(body, &queryRequest)
+		}
+
+		if err != nil {
+			return nil, err
```

Comment:
Question on user input validation on API level - what is the expected behaviour of playground on `/query` endpoint, in case `"limit":"20"` is specified, in payload, instead of `"limit":20` ?

As an API user, I would expect it to result in
```json
{
    "status": 4XX,
    "errors": [
        {
            "status": 4XX,
            "title": "Error description"
        }
    ]
}
```

but I belive current implementation would result in 500 without any payload in response.

Reply:
Improved in 99af6d0. Now limit is optional, and if the json unmarshall fails the reply will contain:
```json
{
    "status": 400,
    "errors": [
        {
            "status": 400,
            "title": "Bad Request. Expected body: { \"query\": \"SQL statement\", \"limit\": 1234 }"
        }
    ]
}
```